clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -527,4 +536,87 @@ public void onComplete(ClientResponse response) {
         boolean shouldBlock();
     }
 
+
+    /*
+     * A threadsafe helper class to hold requests per node that has not been sent yet
+     */
+    private final static class UnsentRequests {
+        private final Map<Node, List<ClientRequest>> unsent;
+
+        public UnsentRequests() {
+            unsent = new HashMap<>();
+        }
+
+        public synchronized void put(Node node, List<ClientRequest> requests) {
+            List<ClientRequest> nodeUnsent = unsent.get(node);
+            if (nodeUnsent == null) {
```

Comment:
In this case, we can we just reuse the received `requests` instead of allocating a new `ArrayList`?